packages/palette/src/elements/Shelf/Shelf.tsx

```diff
@@ -0,0 +1,307 @@
+import { themeGet } from "@styled-system/theme-get"
+import React, {
+  Children,
+  createRef,
+  isValidElement,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react"
+import styled from "styled-components"
+import { useCursor } from "use-cursor"
+import { visuallyDisableScrollbar } from "../../helpers/visuallyDisableScrollbar"
+import { ChevronIcon } from "../../svgs"
+import { Box, BoxProps } from "../Box"
+import { CELL_GAP_PADDING_AMOUNT, paginateCarousel } from "../Carousel"
+import { CarouselBar } from "../CarouselBar"
+import { Clickable } from "../Clickable"
+import { FullBleed } from "../FullBleed"
+
+/** ShelfProps */
+export type ShelfProps = BoxProps & {
+  snap?: "none" | "start" | "end" | "center"
+  children: JSX.Element | JSX.Element[]
+  onChange?(index: number): void
+}
+
+/**
+ * A Shelf is a new kind of carousel...
+ */
+export const Shelf: React.FC<ShelfProps> = ({
+  children,
+  snap = "none",
+  onChange,
+  ...rest
+}) => {
+  const cells = useMemo(
+    () =>
+      Children.toArray(children)
+        .filter(isValidElement)
+        .map((child) => ({ child, ref: createRef<HTMLLIElement>() })),
+    [children]
+  )
+
+  const containerRef = useRef<HTMLDivElement | null>(null)
+  const viewportRef = useRef<HTMLDivElement | null>(null)
+
+  const [mounted, setMounted] = useState(false)
+  const [pages, setPages] = useState([0])
+  const [progress, setProgress] = useState(0)
+  const [offset, setOffset] = useState(0)
+
+  const init = useCallback(() => {
+    if (containerRef.current === null) return
+
+    const { current: container } = containerRef
+
+    // Set page-stops
+    const values = cells.map(({ ref }, i) => {
+      // If we have an offset we actually want to subtract it from
+      // the first and last elements.
+      if (offset !== 0 && (i === 0 || i === cells.length - 1)) {
+        return Math.ceil(ref.current.clientWidth - offset)
+      }
+
+      return ref.current.clientWidth
+    })
+
+    setPages(
+      paginateCarousel({
+        // Here we use the container width instead of the viewport width.
+        // The viewport has been extended to the full width of the window;
+        // we want to scroll to the parent boundaries instead.
+        viewport: container.clientWidth,
+        values,
+      })
+    )
+
+    // Set offset to accomodate full-bleed and line up initially with page-margins
+    const { x } = container.getBoundingClientRect()
+    setOffset(x)
+
+    setMounted(true)
+  }, [cells, offset])
+
+  useEffect(() => {
+    init()
+
+    window.addEventListener("resize", init)
+    return () => {
+      window.removeEventListener("resize", init)
+    }
+  }, [init])
```

Comment:
I had this mostly correct; but the offset value was stale without realizing it. (I had initially only had `cells` in this dependency array — having `init` without the `useCallback` was causing infinite renders).

Note to self that frequently just including some kind of debugging visual really clarifies what's going on — in this case I added markers overlaying where the page stops were and I immediately noticed they were out of sync once I resized the browser.

Reply:
> having init without the useCallback was causing infinite renders

Wow, this makes me think that the whole reason we couldn't do URL driven state updates in thet Artwork filter is that that part of the code wasn't wrapped within a `useCallback` 😠 I remember the same thing happening. Something was being triggered leading to infinite renders. (In any case, with this new artwork filter work I plan on rectifying this.)